src/lib.rs

```diff
@@ -1,109 +1,339 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
-/// A FRAME pallet template with necessary imports
-
-/// Feel free to remove or edit this file as needed.
-/// If you change the name of this file, make sure to update its references in runtime/src/lib.rs
-/// If you remove this file, you can remove those references
-
-/// For more guidance on Substrate FRAME, see the example pallet
-/// https://github.com/paritytech/substrate/blob/master/frame/example/src/lib.rs
-
-use frame_support::{decl_module, decl_storage, decl_event, decl_error, dispatch};
+use codec::{Decode, Encode};
+use fixed::types::U16F16;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch, ensure, sp_runtime::RuntimeDebug,
+    sp_std::prelude::*,
+};
 use frame_system::{self as system, ensure_signed};
+use product_registry::ProductId;
 
 #[cfg(test)]
 mod mock;
 
 #[cfg(test)]
 mod tests;
 
-/// The pallet's configuration trait.
-pub trait Trait: system::Trait {
-	// Add other types and constants required to configure this pallet.
+// General constraints to limit data size
+// Note: these could also be passed as trait config parameters
+pub const IDENTIFIER_MAX_LENGTH: usize = 10;
+pub const SHIPMENT_MAX_PRODUCTS: usize = 10;
+
+// Custom types
+pub type Identifier = Vec<u8>;
```

Comment:
We could use a hash `[u8; 32]` instead of `Vec<u8>`. It is possible to hash the info in JS and keep it more simple to handle here.

Reply:
Indeed. Though the idea here is to store ProductId & ShipmentId on-chain in Shipment & ShippingEvent structs. Using an upper bound on those identifiers (e.g. IDENTIFIER_MAX_LENGTH), they will be smaller than a hash (32 bytes). When used as storage keys in maps, they will be hashed with blake2_128_concat.

Reply:
Right, this make sense. It is a nice optimization. Couldn't we achieve the same if we use an array limited by the same IDENTIFIER_MAX_LENGTH constant. I'm just saying because it makes more sense to let the compiler handle the size and limits. It will also help because it implements the Copy trait which doesn't exists for `Vec<u8>`

Reply:
Makes sense. I had a look but this is not a 1 or 2 line(s) change, so I'll tackle this in a separate issue. Meanwhile I propose to LGTM.

Reply:
Agree, it is not as simple as replacing Vec. I propose to merge this and follow up with another PR.